### PR TITLE
Limiting maximum batch size bytes and adding configuration class

### PR DESCRIPTION
**Pull request by dannycranmer**
*Description of changes:*
Limiting maximum batch size bytes inline with Firehose quota for the given region.
Abstracted Producer properties and validation into a separate immutable class with builder.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java [hunk 1]
```diff
@@ -227,17 +231,86 @@ public void testFirehoseProducerSingleThreadFailedToSendRecords() throws Excepti
...
+    public void testFirehoseProducerBatchesRecordsWithCustomBatchSize() throws Exception {
+        Properties config = new Properties();
+        config.setProperty(FIREHOSE_PRODUCER_BUFFER_MAX_BATCH_BYTES, "100");
+
+        FirehoseProducer<UserRecordResult, Record> producer = createFirehoseProducer(config);
+
+        when(firehoseClient.putRecordBatch(any(PutRecordBatchRequest.class)))
+                .thenReturn(new PutRecordBatchResult());
+
+        // Fill up the maximum capacity: 8 * 512kB = 4MB
```
Comment (GS-KeithLee) [inline src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java:272]: Comment on code needs to be changed

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java [hunk 2]
```diff
@@ -227,17 +231,86 @@ public void testFirehoseProducerSingleThreadFailedToSendRecords() throws Excepti
...
+        IntStream.range(0, 2).forEach(i -> addRecord(producer, 100));
+
+        Thread.sleep(2000);
+
+        assertThat(firehoseProducer.getOutstandingRecordsCount()).isEqualTo(0);
+        verify(firehoseClient, times(2)).putRecordBatch(putRecordCaptor.capture());
+
+        // The first batch should contain 4 records (up to 4MB), the second should contain the remaining record
```
Comment (GS-KeithLee) [inline src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java:280]: Comment on code needs to be changed

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java (hunk 1 above)
Comment (GS-KeithLee) [reply src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java:272]: Speficially to 2 * 100

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java (hunk 2 above)
Comment (GS-KeithLee) [reply src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java:280]: Also to to 2 * 100

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java (hunk 1 above)
Comment (dannycranmer) [reply src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java]: Fixed

File: src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java (hunk 2 above)
Comment (dannycranmer) [reply src/test/java/com/amazonaws/services/kinesisanalytics/flink/connectors/producer/impl/FirehoseProducerTest.java]: Fixed

**Pull request merged**